common-ui-compose/src/main/java/app/tivi/common/compose/Insets.kt

```diff
@@ -0,0 +1,300 @@
+ * limitations under the License.
+ */
+
+@file:Suppress("NOTHING_TO_INLINE")
+
+package app.tivi.common.compose
+
+import androidx.compose.Composable
+import androidx.compose.Providers
+import androidx.compose.Stable
+import androidx.compose.StructurallyEqual
+import androidx.compose.getValue
+import androidx.compose.mutableStateOf
+import androidx.compose.onCommit
+import androidx.compose.remember
+import androidx.compose.setValue
+import androidx.compose.staticAmbientOf
+import androidx.core.view.ViewCompat
+import androidx.core.view.doOnAttach
+import androidx.ui.core.DensityAmbient
+import androidx.ui.core.Modifier
+import androidx.ui.core.ViewAmbient
+import androidx.ui.core.composed
+import androidx.ui.layout.absolutePadding
+import androidx.ui.layout.height
+import androidx.ui.layout.width
+import androidx.ui.unit.Density
+import androidx.ui.unit.dp
+
+/**
+ * Main holder of our inset values.
+ */
+@Stable
+class DisplayInsets {
+    val systemBars by lazy(LazyThreadSafetyMode.NONE) { Insets() }
+}
+
+@Stable
+class Insets {
+    var left by mutableStateOf(0.dp, StructurallyEqual)
+        internal set
+    var top by mutableStateOf(0.dp, StructurallyEqual)
+        internal set
+    var right by mutableStateOf(0.dp, StructurallyEqual)
+        internal set
+    var bottom by mutableStateOf(0.dp, StructurallyEqual)
+        internal set
+
+    /**
+     * TODO: doesn't currently work
+     */
+    var visible by mutableStateOf(true)
+        internal set
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+        other as Insets
+        if (left != other.left) return false
+        if (top != other.top) return false
+        if (right != other.right) return false
+        if (bottom != other.bottom) return false
+        if (visible != other.visible) return false
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = left.hashCode()
+        result = 31 * result + top.hashCode()
+        result = 31 * result + right.hashCode()
+        result = 31 * result + bottom.hashCode()
+        result = 31 * result + visible.hashCode()
+        return result
+    }
+}
+
+val InsetsAmbient = staticAmbientOf<DisplayInsets>()
+
+@Composable
+fun ProvideInsets(content: @Composable () -> Unit) {
+    val view = ViewAmbient.current
+    val density = DensityAmbient.current
+
+    val displayInsets = remember { DisplayInsets() }
+
+    onCommit(view) {
+        ViewCompat.setOnApplyWindowInsetsListener(view) { _, windowInsets ->
+            displayInsets.systemBars.updateFrom(windowInsets.systemWindowInsets, density)
+
+            // Return the unconsumed insets
+            windowInsets
+        }
+        view.doOnAttach { it.requestApplyInsets() }
+
+        onDispose {
+            ViewCompat.setOnApplyWindowInsetsListener(view, null)
+        }
+    }
+
+    Providers(InsetsAmbient provides displayInsets) {
```

Comment:
Consider factoring out a producer function that just returns the remembered and managed DisplayInsets, so that this abstraction doesn't have to be responsible for managing a Providers call and content function.

Reply:
I tried, but I'm not sure this is going to work as-is. A view can only have one `OnApplyWindowInsetsListener` set, so if the producer function (say called `displayInsets()`) is called in multiple places, only one them will actually work.

I could shove a multiplexing `OnApplyWindowInsetsListener` or a `InsetsController` to an Ambient, and then have each `displayInsets()` subscribe to it. WDYT?